### PR TITLE
feat(cardinal): implement VolatileStorage interface

### PR DESCRIPTION
**Pull request by pyrofolium**
Closes: WORLD-812

This is a design proposal. 

For ECB there shall now be two interfaces. One for Volatile storage, and the other for external IO storage. 

Redis will be placed behind the IO storage (This is just PrimitiveStorage nothing has changed) Technically redis is volatile too but whatever. 

All maps under ECB will be replaced by VolatileStorage which is VolatileStorage[K, V] and type safe across a generic K and a generic V. This replaces the design where everything (maps and redis) was going to be replaced by PrimitiveStorage. 

Basically I just took all the design inputs from everyone and just combined it into one. 

This PR only implements VolatileStorage for two properties on ECB, it is incomplete, but your evaluation here requires you to consider replacing all properties involving map with VolatileStorage as this PR represents a design proposal that IF approved will represent the beginnings of fully decoupling ECB from maps.

### CODE REVIEW

**Comment by pyrofolium** [comment]
Alternative implimentation:
```go
type VolatileStorage[K comparable] interface {
        get(key K) (any, error)
}

func Get[K, V] (storage VolatileStorage[K], key K) (V, error) {
        res, err := get(k)
        if err != nil {
               return err
        }
        
}
```

Or something along these lines. The above though not type safe allows us to consolidate storage into a single instance. Just an alternative direction here.

**Comment by jerargus** [comment]
Replacing ECB’s volatile workspace with an interface does not seem useful to me. The main pro I’ve heard is that we can replace it with a more efficient/faster implementation. 

1. Is this actually a bottleneck? Our performance tests (and intuition) points to serialization/deserilization, writing to permanent storage, and network calls. This volatile storage is separate from all of those concerns.
2. What alternatives are we considering? The one posted in the ticket (concurrent-swiss-map) says it’s faster than sync.Map. We don’t use sync map. Access to volatile storage is not happening concurrently.

I don’t see a good reason to replace built-in maps here. Using built in maps make our code more maintainable and clear, especially to new members of the team.

I don’t think replacing volatile storage here should be a priority. The priority should be trying different libraries for IO storage and benchmarking/comparing them.

**Comment by pyrofolium** [comment]
> Replacing ECB’s volatile workspace with an interface does not seem useful to me. The main pro I’ve heard is that we can replace it with a more efficient/faster implementation.
> 
> 1. Is this actually a bottleneck? Our performance tests (and intuition) points to serialization/deserilization, writing to permanent storage, and network calls. This volatile storage is separate from all of those concerns.
> 2. What alternatives are we considering? The one posted in the ticket (concurrent-swiss-map) says it’s faster than sync.Map. We don’t use sync map. Access to volatile storage is not happening concurrently.
> 
> I don’t see a good reason to replace built-in maps here. Using built in maps make our code more maintainable and clear, especially to new members of the team.
> 
> I don’t think replacing volatile storage here should be a priority. The priority should be trying different libraries for IO storage and benchmarking/comparing them.

This statement is literally at odds with this where you approved of the initial interface on top of maps: 

https://github.com/Argus-Labs/world-engine/pull/619#pullrequestreview-1898724044

All I did was change the interface signature. 

All three of us agreed to a compromise That's me, scott @smsunarto and you @jerargus during a conversation right before scott left for the airport. 

Now you're saying you don't approve of the interface all together? This PR was meant ONLY meant for consideration of the interface signature, which wasn't fully agreed upon yet. It seems like you're making contradictory statements. 

Not to be argumentative here @jerargus but things like this are literally the problem I'm having with a lot of your PRs. I feel like I'm suddenly facing hard blocks out of nowhere... we should talk about this on Thurs. I'm sure this is not what's going on but we should resolve this in person. 

Anyway if you changed your mind, that's fine, but please indicate such, as contradictory statements make it seem odd and unclear as to what your intentions are. No offense intended here jer, just telling the situation as it looks like from my perspective.

**Comment by pyrofolium** [comment]
> 1. Is this actually a bottleneck? Our performance tests (and intuition) points to serialization/deserilization, writing to permanent storage, and network calls. This volatile storage is separate from all of those concerns.

No it's not a big bottleneck. But it doesn't hurt to decouple this now before ECB is upgraded to V2 otherwise it's a bigger untangling of Code for WHEN we want to replace it. Documentation on ECS indicate that the way the data is laid out it here is extremely important so when all other bottlenecks like http and redis are eliminated this will inevitably be a target for optimization as all other ECS frameworks basically optimize this area with Data Oriented designs. 

The optimization doesn't have to happen any time soon. ECS V1 -> V2 will complicate things a lot as it converts ECB into a linked list. It's better to do a trivial decoupling now so that we don't have to untangle it later. 

> 2. What alternatives are we considering? The one posted in the ticket (concurrent-swiss-map) says it’s faster than sync.Map. We don’t use sync map. Access to volatile storage is not happening concurrently.

That's true, but in order to get rid of redis we will have to switch over concurrent read writes on volatile storage. Redis will become just a back up snapshotter that is saved to concurrently but not on every tick. So with an interface we can try both. Sync.Map and concurrent-swiss-map. 

We discussed this all already right before Scott left to the airport. Happy to discuss all of this again if your opinion has changed. 

Let's talk thurs. Looks like this will be blocked for until After Scott gets back.

**Review by jerargus — APPROVED** [review]

**Comment by pyrofolium** [comment]
This needs re-review. I updated the PR to replace every map store in ECB with an interface. This ups the size of the PR by quite a bit. 

No other change was done.

File: cardinal/gamestate/map.go [hunk 1]
```diff
@@ -18,257 +16,37 @@ func NewMapStorage[K comparable, V any]() *MapStorage[K, V] {
...
-		return v - 1, true
-	case float64:
-		return v - 1, true
-	case complex64:
-		return v - complex(1, 0), true
-	case complex128:
-		return v - complex(1, 0), true
-	default:
-		return nil, false
+		return v, eris.New("key not found")
```
Comment (jerargus) [inline cardinal/gamestate/map.go:35]: This interface should distinguish between actual errors and missing keys. "not found" could either be a dedicated error, or Get could return `(V, bool, error)`

File: cardinal/gamestate/read_only.go [hunk 2]
```diff
@@ -146,9 +154,11 @@ func (r *readOnlyManager) SearchFrom(filter filter.ComponentFilter, start int) *
 	if err := r.refreshArchIDToCompTypes(); err != nil {
 		return itr
 	}
-	for i := start; i < len(r.archIDToComps); i++ {
+	for i := start; i < r.archIDToComps.Len(); i++ {
 		archID := types.ArchetypeID(i)
-		if !filter.MatchesComponents(types.ConvertComponentMetadatasToComponents(r.archIDToComps[archID])) {
+		//TODO: error swallowed here.
```
Comment (jerargus) [inline cardinal/gamestate/read_only.go:159]: This TODO should be tied to a ticket in our backlog.

**Review by namrapatel — COMMENTED** [review]
We should also clean the PR title to be, "feat(cardinal): implement VolatileStorage interface"

File: cardinal/gamestate/ecb.go [hunk 3]
```diff
@@ -72,34 +72,52 @@ func NewEntityCommandBuffer(storage PrimitiveStorage[string]) (*EntityCommandBuf
...
+	err = m.activeEntities.Clear()
+	if err != nil {
+		return err
+	}
+	ids, err := m.entityIDToOriginArchID.Keys()
+	if err != nil {
+		return err
+	}
+	for _, id := range ids {
+		err = m.entityIDToArchID.Delete(id)
```
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/gamestate/ecb.go:103]: I'm just dumbly following the original code. The original code is not clearing all the ids. It gets ids from another source. So it's not guaranteed that all ids will be deleted.

File: cardinal/gamestate/ecb.go [hunk 4]
```diff
@@ -72,34 +72,52 @@ func NewEntityCommandBuffer(storage PrimitiveStorage[string]) (*EntityCommandBuf
...
+		return err
 	}
-	clear(m.entityIDToOriginArchID)
 
 	m.isEntityIDLoaded = false
 	m.pendingEntityIDs = 0
 
 	for _, archID := range m.pendingArchIDs {
-		delete(m.archIDToComps, archID)
+		err = m.archIDToComps.Delete(archID)
```
[the remark this answers is not in the document]
Comment (pyrofolium) [reply cardinal/gamestate/ecb.go:117]: pendingArchIDs is not guaranteed to cover every ID in m.archIDToComps, just like above. Again I'm just dumbly following the logic of the original code. They did it that way, so I do it the same way. No difference.
